libnopegl/src/glsl/distmap.frag

```diff
@@ -0,0 +1,506 @@
+        vec2 off4 = c_div(d4, vec2(1.0, 0.0) - c_mul(d4, sum_of_inv(prv4, prv0, prv1, prv2, prv3)));
+
+        prv0 -= off0;
+        prv1 -= off1;
+        prv2 -= off2;
+        prv3 -= off3;
+        prv4 -= off4;
+
+        if (SMALL_OFF(off0) && SMALL_OFF(off1) && SMALL_OFF(off2) && SMALL_OFF(off3) && SMALL_OFF(off4))
+            break;
+    }
+
+    /* Keep only the real values */
+    int nroot = 0;
+    if (close_to_zero(prv0.y)) roots[nroot++] = prv0.x;
+    if (close_to_zero(prv1.y)) roots[nroot++] = prv1.x;
+    if (close_to_zero(prv2.y)) roots[nroot++] = prv2.x;
+    if (close_to_zero(prv3.y)) roots[nroot++] = prv3.x;
+    if (close_to_zero(prv4.y)) roots[nroot++] = prv4.x;
+    return nroot;
+}
+
+/* Linear: f(x)=ax+b */
+int root_find1(out float root, float a, float b)
+{
+    if (close_to_zero(a))
+        return 0;
+    root = -b / a;
+    return 1;
+}
+
+/* Quadratic: f(x)=ax²+bx+c */
+int root_find2(out vec2 roots, float a, float b, float c)
+{
+    if (close_to_zero(a))
+        return root_find1(roots.x, b, c);
+
+    float m = -b / (2.0 * a);
+    float delta = m*m - c / a;
+    if (close_to_zero(delta)) {
+        roots.x = m;
+        return 1;
+    }
+    if (delta < 0.0)
+        return 0;
+    float z = sqrt(delta);
+    roots = m + vec2(-z, z);
+    return 2;
+}
+
+/* Cubic: f(x)=ax³+bx²+cx+d */
+int root_find3(out vec3 roots, float a, float b, float c, float d)
+{
+    if (close_to_zero(a))
+        return root_find2(roots.xy, b, c, d);
+    return aberth_ehrlich_3(roots, a, b, c, d);
+}
+
+/* Quartic: f(x)=ax⁴+bx³+cx²+dx+e */
+int root_find4(out vec4 roots, float a, float b, float c, float d, float e)
+{
+    if (close_to_zero(a))
+        return root_find3(roots.xyz, b, c, d, e);
+    return aberth_ehrlich_4(roots, a, b, c, d, e);
+}
+
+/* Quintic: f(x)=ax⁵+bx⁴+cx³+dx²+ex+f */
+int root_find5(out float roots[5], float a, float b, float c, float d, float e, float f)
+{
+    if (close_to_zero(a)) {
+        vec4 roots4 = vec4(0.0);
+        int nroot = root_find4(roots4, b, c, d, e, f);
+        roots = float[5](roots4.x, roots4.y, roots4.z, roots4.w, 0.0);
+        return nroot;
+    }
+    return aberth_ehrlich_5(roots, a, b, c, d, e, f);
+}
+
+vec3 root_find3_expected(int topology, float a, float b, float c, float d)
+{
+    vec3 r = vec3(LARGE_FLOAT);
+    int n = root_find3(r, a, b, c, d);
+    if (n == 3) {
+        if (r.x > r.y) r.xy = r.yx;
+        if (r.x > r.z) r.xz = r.zx;
+        if (r.y > r.z) r.yz = r.zy;
+        return stitch_3_roots(topology, r.xyz, a, b, c);
+    }
+    if (n == 2) {
+        if (r.x > r.y) r.xy = r.yx;
+        return stitch_2_roots(topology, r.xy, a, b, c);
+    }
+    if (n == 1) return stitch_1_root(topology, r.x, a, b, c);
+    return vec3(LARGE_FLOAT); // cannot return r because it might have been tainted by root_find3()
+}
+
+vec4 get_color(vec2 p)
+{
+    // TODO: can we reduce the number of polynomial to evaluate for a given pixel?
+
```

Comment:
spurious newline ?

Reply:
Not really, it's a comment for the whole double loops. I moved the comment above the function prototype.